### PR TITLE
Windows support [part 5] 

### PR DESCRIPTION
**Pull request by petrutlucian94**
Windows support [part 5]

This PR series intends to set the ground for the upcoming Windows port effort.
For now, we're mostly interested in the client side, allowing Windows hosts
to consume rados, rbd and cephfs resources.

Those PRs should be reviewed and merged in order. Each PR depends on the previous one from the series (in fact it includes the previous commits as well).

~~Part 1: #31981~~
~~Part 2: #32704~~
~~Part 3: #32705~~
~~Part 4: #32707~~
**Part 5: #32780** 
Part 6: #32779
Part 7: #32778 
Part 8: #32777 
Part 9: #32776
Part 10: #33750
Part 11: #34858
Part 12: #34859

Related but independent PRs:
https://github.com/ceph/ceph/pull/32027
https://github.com/ceph/fmt/pull/2
https://github.com/ceph/rocksdb/pull/42

For more details about the build process and current status, please check the
README.windows.rst file. Worth mentioning that the resulted binaries can
connect to ceph clusters and consume pools: http://paste.openstack.org/raw/787534/

Here are some test results as well:
[test_results.zip](https://github.com/ceph/ceph/files/4077517/test_results.zip)

Individual PR commits:

```
rbd: make libcurses optional
common: add Windows implementation for system info getter
common: add platform checks
common: skip undefined open flags on Windows
os: Add missing include
```

Any feedback is highly appreciated.

Signed-off-by: Lucian Petrut lpetrut@cloudbasesolutions.com
Signed-off-by: Alin Gabriel Serdean aserdean@cloudbasesolutions.com

### CODE REVIEW

**Comment by petrutlucian94** [comment]
jenkins test make check

**Comment by petrutlucian94** [comment]
jenkins test make check

**Comment by petrutlucian94** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
@petrutlucian94 could you rebase against the latest master HEAD?

File: src/os/filestore/os_xattr.h [hunk 1]
```diff
@@ -24,6 +24,10 @@ extern "C" {
 # define ENOATTR ENODATA
 #endif
 
+#ifdef _WIN32
+#include <crtdefs.h>
```
Comment (tchaikov) [inline src/os/filestore/os_xattr.h:28]: i assume that the libc used by MinGW also complies to at least C89/C90. so per https://en.cppreference.com/w/c/types/size_t, `stddef.h` should suffice.

Comment (petrutlucian94) [reply src/os/filestore/os_xattr.h:28]: Yep, that's right. I'll use stddef.h instead.

File: src/common/util.cc [hunk 2]
```diff
@@ -63,6 +67,24 @@ int get_fs_stats(ceph_data_stats_t &stats, const char *path)
...
+  if (!(GetDiskFreeSpaceExA(path, &avail_bytes,
+                            &total_bytes, &total_free_bytes))) {
+    return -EINVAL;
+  }
+
+  stats.byte_total = total_bytes.QuadPart;
+  stats.byte_used = total_bytes.QuadPart - total_free_bytes.QuadPart;
+  // may not be equal to total_free_bytes due to quotas
+  stats.byte_avail = avail_bytes.QuadPart;
+  stats.avail_percent = (((float)stats.byte_avail/stats.byte_total)*100);
```
Comment (tchaikov) [inline src/common/util.cc:84]: please add spaces around operator, and might want to remove out-most parentheses .

File: src/common/util.cc [hunk 3]
```diff
@@ -63,6 +67,24 @@ int get_fs_stats(ceph_data_stats_t &stats, const char *path)
   stats.avail_percent = (((float)stats.byte_avail/stats.byte_total)*100);
   return 0;
 }
+#else
+int get_fs_stats(ceph_data_stats_t &stats, const char *path)
+{
+  ULARGE_INTEGER avail_bytes, total_bytes, total_free_bytes;
+
+  if (!(GetDiskFreeSpaceExA(path, &avail_bytes,
```
Comment (tchaikov) [inline src/common/util.cc:75]: might want to drop the parentheses around the function call.

File: src/common/util.cc [hunk 4]
```diff
@@ -172,8 +195,37 @@ int get_cgroup_memory_limit(uint64_t *limit)
...
   return ret;
+  #else
+  return 0;
+  #endif
 }
 
+#ifdef _WIN32
+int get_windows_version(POSVERSIONINFOEXW ver) {
+    HMODULE hLib;
+    int ret_val = 0;
```
Comment (tchaikov) [inline src/common/util.cc:206]: could you move the definition of variables closer to where they are used for the first time for better readability?

File: src/common/util.cc [hunk 5]
```diff
@@ -172,8 +195,37 @@ int get_cgroup_memory_limit(uint64_t *limit)
...
+
+    typedef DWORD (WINAPI *RtlGetVersion_t) (OSVERSIONINFOEXW*);
```
Comment (tchaikov) [inline src/common/util.cc:208]:
```suggestion
    using  get_version_func_t = DWORD (WINAPI *)(OSVERSIONINFOEXW*);
```
for better readability.

File: src/common/util.cc [hunk 6]
```diff
@@ -172,8 +195,37 @@ int get_cgroup_memory_limit(uint64_t *limit)
...
   fclose(f);
   return ret;
+  #else
+  return 0;
+  #endif
 }
 
+#ifdef _WIN32
+int get_windows_version(POSVERSIONINFOEXW ver) {
+    HMODULE hLib;
```
Comment (tchaikov) [inline src/common/util.cc:205]: please avoid using hungarian notation.
```suggestion
    HMODULE ntdll_lib;
```

File: src/common/util.cc (hunk 4 above)
Comment (tchaikov) [inline src/common/util.cc:206]:
```suggestion
    int ret = 0;
```

File: src/common/util.cc [hunk 7]
```diff
@@ -192,6 +245,44 @@ void collect_sys_info(map<string, string> *m, CephContext *cct)
...
+  OSVERSIONINFOEXW ver = {0};
+  ver.dwOSVersionInfoSize = sizeof(ver);
+  get_windows_version(&ver);
+
+  char version_str[64];
+  snprintf(version_str, 64, "%d.%d (%d)",
+           ver.dwMajorVersion, ver.dwMinorVersion, ver.dwBuildNumber);
+
+  char hostname[64];
+  DWORD hostname_sz = 64;
```
Comment (tchaikov) [inline src/common/util.cc:258]:
```suggestion
  DWORD hostname_sz = sizeof(hostname);
```

File: src/tools/rbd/CMakeLists.txt [hunk 8]
```diff
@@ -51,15 +52,22 @@ set(rbd_srcs
   action/Trash.cc
   action/Watch.cc)
 
+if(CURSES_FOUND)
+  add_definitions(-DHAVE_CURSES)
```
Comment (tchaikov) [inline src/tools/rbd/CMakeLists.txt:56]: please don't add definitions globally if not necessary. might want to use `target_compile_definitions()` instead.

File: src/tools/rbd/action/Perf.cc [hunk 9]
```diff
@@ -560,6 +563,7 @@ class MainWindow {
 };
 
 } // namespace iotop
+#endif /* HAVE_CURSES */
```
Comment (tchaikov) [inline src/tools/rbd/action/Perf.cc:566]: nit, 
```suggestion
#endif // HAVE_CURSES
```

**Comment by petrutlucian94** [comment]
> could you prefix the title of your commit message with the subcomponent your are changing ? see https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#3-describe-your-changes

Will do. Thanks a lot for reviewing this PR. I'm in vacation for the next two weeks but I'll try to find some time to address the comments.

File: src/tools/rbd/action/Perf.cc [hunk 10]
```diff
@@ -688,13 +693,15 @@ int execute_iotop(const po::variables_map &vm,
...
 }
 
-Shell::Action stat_action(
-  {"perf", "image", "iostat"}, {}, "Display image IO statistics.", "",
-  &get_arguments_iostat, &execute_iostat);
 Shell::Action top_action(
   {"perf", "image", "iotop"}, {}, "Display a top-like IO monitor.", "",
   &get_arguments_iotop, &execute_iotop);
 
+#endif /* HAVE_CURSES */
```
Comment (dillaman) [inline src/tools/rbd/action/Perf.cc:700]: Nit: same comment here

File: src/osdc/Objecter.h [hunk 11]
```diff
@@ -28,6 +28,8 @@
 
 #include <boost/container/small_vector.hpp>
 #include <boost/asio.hpp>
+#include <boost/thread/shared_mutex.hpp>
```
Comment (dillaman) [inline src/osdc/Objecter.h:31]: Nit: I don't see these being used in this header.

Comment (petrutlucian94) [reply src/osdc/Objecter.h]: Right, looks like it's no longer needed.

File: src/common/util.cc [hunk 12]
```diff
@@ -145,6 +167,7 @@ static void distro_detect(map<string, string> *m, CephContext *cct)
 
 int get_cgroup_memory_limit(uint64_t *limit)
 {
+  #ifndef _WIN32
```
Comment (tchaikov) [inline src/common/util.cc:170]: i'd prefer avoid indent unless necessary, for instance it's nested in yet another `#ifdef`.

File: src/common/util.cc [hunk 13]
```diff
@@ -172,8 +195,35 @@ int get_cgroup_memory_limit(uint64_t *limit)
 out:
   fclose(f);
   return ret;
+  #else
```
Comment (tchaikov) [inline src/common/util.cc:198]: ditto.

File: src/common/util.cc [hunk 14]
```diff
@@ -172,8 +195,35 @@ int get_cgroup_memory_limit(uint64_t *limit)
...
+  return 0;
+  #endif
```
Comment (tchaikov) [inline src/common/util.cc:200]: ditto.

File: src/common/util.cc [hunk 15]
```diff
@@ -192,6 +243,44 @@ void collect_sys_info(map<string, string> *m, CephContext *cct)
...
+
+  char hostname[64];
+  DWORD hostname_sz = sizeof(hostname);
+  GetComputerNameA(hostname, &hostname_sz);
+
+  SYSTEM_INFO sys_info;
+  char* arch_str;
+  GetNativeSystemInfo(&sys_info);
+
+  switch(sys_info.wProcessorArchitecture) {
```
Comment (tchaikov) [inline src/common/util.cc:263]: please add a space after `switch`.

File: src/tools/rbd/CMakeLists.txt [hunk 16]
```diff
@@ -61,8 +62,12 @@ target_link_libraries(rbd librbd
   journal
   libneorados
   librados
-  ceph-common global ${CURSES_LIBRARIES}
+  ceph-common global
   ${BLKID_LIBRARIES} ${CMAKE_DL_LIBS})
+if(CURSES_FOUND)
+  target_compile_definitions(rbd PRIVATE HAVE_CURSES)
```
Comment (tchaikov) [inline src/tools/rbd/CMakeLists.txt:68]: i don't think this is needed.

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2020-07-30_15:58:21-rados-wip-kefu-testing-2020-07-30-2107-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-07-31_01:42:48-rados-wip-kefu-testing-2020-07-30-2107-distro-basic-smithi/

failure tracked by

- https://tracker.ceph.com/issues/46323

**Pull request merged**